nixos/modules/services/networking/networkmanager.nix

```diff
@@ -379,6 +380,52 @@ in {
           for more details.
         '';
       };
+      ensure-profiles = {
+        profiles = mkOption {
+          type = types.attrsOf ini.type;
+          default = { };
+          example = {
+            home-wifi = {
+              connection = {
+                id = "home-wifi";
+                type = "wifi";
+                permissions = "";
+              };
+              wifi = {
+                mac-address-blacklist = "";
+                mode = "infrastructure";
+                ssid = "Home Wi-Fi";
+              };
+              wifi-security = {
+                auth-alg = "open";
+                key-mgmt = "wpa-psk";
+                psk = "$PASSWORD";
+              };
+              ipv4 = {
+                dns-search = "";
+                method = "auto";
+              };
+              ipv6 = {
+                addr-gen-mode = "stable-privacy";
+                dns-search = "";
+                method = "auto";
+              };
+            };
+          };
+          description = lib.mdDoc ''
+            declaratively define NetworkManager profiles.
+          '';
```

Comment:
Can you maybe give a link to some documentation that describes valid fields here?

Reply:
Also you should explain that this will only create the profiles in memory and never delete them.

Question: What happens when the user changes a profile manually after it got created with ensureProfiles? Will the profile and the change get persisted? Will the next nixos-rebuild overwrite the change? Is the profile read only? If its readonly do networkmanager clients have an acceptable UI for that?

Reply:
> Can you maybe give a link to some documentation that describes valid fields here?

I think I'll just add a link to redhats docs if that is fine? 
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/assembly_networkmanager-connection-profiles-in-keyfile-format_configuring-and-managing-networking
Other then that you can use your current profiles that are most likely in `/etc/NetworkManager/system-connections` and you can use github.com/janik-haag/nm2nix to convert those to the needed nix code. 

> Question: What happens when the user changes a profile manually after it got created with ensureProfiles? 

At first nothing (only applies for edits with editors like vim tools like nmtui follow later) because network-manager needs a reload with `nmcli connection reload` to scan them and yes then they would work but only as long as `NetworkManager-ensure-profiles` doesn't run again because it [just overrides files](https://github.com/NixOS/nixpkgs/pull/254647/files#diff-0a708e7b053cf5df7620b5262936553af2242d2ce9dabde5bbeba221ece0a021R552). The unit will run at every boot and on activation so a nixos-rebuild will also trigger it. 

To be clear if a user manually adds a profile with tools like nm-applet or the gnome one they will be stored in `/etc/NetworkManager/system-connections` and won't be touched by this tool so it is possible to use this in combination with imperative/adhoc managment and at least nm-applet just displays all profiles even when there are name collisions (I didn't test this with the gnome gui or any other). 

> If its readonly do networkmanager clients have an acceptable UI for that?

More or less, if a readonly file in temp gets edited with nmtui or nm-applet they just make the file writable and move it to the persistent directory which shouldn't bother us too much.

Reply:
I'll add this in a structured form to the description

Reply:
Ah, I was not thinking about what happens if someone manually edits the /run files. I meant the nmtui edit.
It will persist the profile in another dir, okay. Does then the persistent profile take precedence over the one created in this profile or the other way around? Either way, I see the danger of very confused users because they can make changes that will have no effect. I think the documentation of this option should warn about that case.

Reply:
> It will persist the profile in another dir,

no it will move the file. So at first there is only one profile with this name because the file was moved, there will be two as soon as the systemd unit runs again because it also creates on in the temp storage again.

> Does then the persistent profile take precedence over the one created in this profile or the other way around?

Nor it just show you two profiles with the same name.

> Either way, I see the danger of very confused users because they can make changes that will have no effect. I think the documentation of this option should warn about that case.

Yup I'll add that.